### PR TITLE
CI: Add Conan testing, mass building and publishing

### LINKED ISSUE #635: [feature] Conan package.
Any chance we can get a conan package? Thank you.

### PR DESCRIPTION
**Pull request by Croydon**
This
  * is adding a recipe for the Conan package manager
  * also adds a small test to the CI if Conan package creation is working 
  * updated: also adds a upload script to the CI for the Conan recipe to a Conan repository

~~Please note that for future releases of benchmark the version string needs to be updated in conanfile.py~~ (not anymore, see discussion)

_______
~~**Future work** could include~~
**Remaining to do list:**
  * [x] setting up a free repository on Bintray: https://bintray.com/dominichamon/benchmark
  * [ ] including the package in the official Conan repository [conan-center](https://bintray.com/conan/conan-center) 
   * [x] and extending the CI to upload new versions of the recipe and built packages to the Bintray repository automatically

To use the recipe after this pull request is merged it requires to clone this git repository first; when the recipe is included in conan-center every Conan client would be able to use it via one single install command.
_______

Please let me know if you have any further questions regarding supporting Conan 😄 

Fixes #635 

//cc @raulbocanegra @p-groarke @danimtb @Mikayex @mpusz @iblis-ms

### CODE REVIEW

**Comment by Croydon** [comment]
The failure of Travis seems to be unrelated since I have neither touched source code nor the previous existing Travis jobs.

File: conanfile.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+from conans import ConanFile, CMake, tools
+
+
+class GoogleBenchmarkConan(ConanFile):
+    name = "benchmark"
+    version = "1.4.1"
```
Comment (dominichamon) [inline conanfile.py]: if this needs to be updated on release, please add it to the releasing.md document.

As an aside, this is very unfortunate. Is there no way to generate this file from cmake passing in the version as a parameter?

File: conanfile.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    def configure(self):
+        if self.settings.os == 'Windows' and self.options.shared:
+            raise Exception("Windows shared builds are not supported right now, see issue #639")
+
+        if self.options.enable_testing == False:
+            self.options.enable_gtest_tests = False
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
+        cmake.definitions['BENCHMARK_ENABLE_TESTING'] = "ON" if self.options.enable_testing else "OFF"
```
Comment (dominichamon) [inline conanfile.py]: if this is enabled, i think you'd also want BENCHMARK_DOWNLOAD_DEPENDENCIES to be ON, or the build won't find googletest.

Comment (dominichamon) [reply conanfile.py]: Ah, just saw the build_requires below. nice.

File: conanfile.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        self.copy(pattern="LICENSE", dst="licenses")
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
+        if self.settings.os == "Linux":
+            self.cpp_info.libs.extend(["pthread", "rt"])
+        elif self.settings.os == "Windows":
```
Comment (dominichamon) [inline conanfile.py]: solaris? you'll need kstat.

File: conanfile.py (hunk 1 above)
Comment (danimtb) [reply conanfile.py]: The version populated from other files such as cmakefiles or headers. Here is the howto: https://docs.conan.io/en/latest/howtos/capture_version.html

File: conanfile.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        "enable_exceptions": [True, False],
+        "enable_lto": [True, False],
+        "enable_testing": [True, False],
+        "enable_gtest_tests": [True, False]
+    }
+    default_options = "shared=False", "fPIC=True", "enable_exceptions=True", "enable_lto=False", "enable_testing=False", "enable_gtest_tests=False"
+    exports_sources = ["*"]
+    generators = "cmake"
+
+    build_subfolder = "."
```
Comment (LebedevRI) [inline conanfile.py]: Does `"."` mean that it will be built in the source tree?
It should be built in some new subdirectory, e.g. `build`.

File: conanfile.py (hunk 1 above)
Comment (LebedevRI) [reply conanfile.py]: Is this version info needed always, or for releases only?
Having it always is actually actively misleading, as this is `1.4.1+git...`
Can it be present for releases only, and otherwise be `version = "unknown"` or something ?

File: conanfile.py (hunk 4 above)
Comment (Croydon) [reply conanfile.py]: Conan is creating working and package directories in it's own Conan directory. The original source origin will never be modified.

Comment (Croydon) [reply conanfile.py]: benchmark

File: conanfile.py (hunk 1 above)
Comment (dominichamon) [reply conanfile.py]: @LebedevRI that's a good point. maybe we can do something like have the default be empty with a check on package creation that it's been set? and then a test that we should run on releases to ensure it's set?

**Comment by Croydon** [comment]
## Tl;dr:
  * we can get rid of the version string 
  * but then I even more strongly advocate for setting up a Conan repository on Bintray than I would have done anyway :)
 
________

## Some thoughts..

A conan reference has the following elements:

`<name> / <version> @ <user> / <channel>`

So the variables for Benchmark are version & channel:

`benchmark / <version> @ google / <channel>`

The version is in most cases the common version number like 1.4.1

The channel is freely selectable, but common practice within the Conan ecosystem is to use either `stable` or `testing`.

The full package reference is decided at the Conan export step. So the reference, including the version and channel can't be changed at a later step (without a re-export(, or some form of alias)).
The channel is not set within the Conan recipe, but rather directly at export, together with the user, like this:

`conan export . <user>/<channel>`
`conan export . google/testing`

So common practice is: Name + version from the recipe; user + channel from the export step.

However, I fully understand that it is annoying to change the version via commit for each release and maybe even change it again directly after the release to something like "1.4.2-dev".

It is possible to remove the version string from the recipe, which I have done in my latest commits. This requires then to specify the full reference at the export step like:

`conan export . benchmark/1.4.1@google/stable`

By this way we got rid of the annoyance for the people doing the releases, but now got an very uncomfortable situation for all consumers of the recipe, which is a strong contradiction to one of the main goals of a package manager: Making it easier to use and handle the library. Consumers would now be required to clone the repository, then doing the export step with a reference which hopefully doesn't bring in some misunderstanding about what Benchmark version is currently in use; also two projects may use completely different references, even though both use the very same version of Benchmark.


## Solution

The annoyance of manually maintaining the version string and the mess and complication for consumers can be best avoided by setting up a Conan repository and let Travis upload new recipe versions automatically.

I have already written the code for this. I have done it in a way that every tag gets a 
`benchmark/<version>@google/stable` release and there is an evergreen `benchmark/master@google/testing` version.

The things you would now have to do would include:
 
  * use/create a free account on http://bintray.com
     * there is already a Google account, not sure if you would like to use this or setup a separate Benchmark one 
  * create a new Conan repository ("Add New Repository" -> Type: Conan)
  * Go to https://travis-ci.org/google/benchmark/settings and create the environment variables CONAN_UPLOAD, CONAN_LOGIN_USERNAME and CONAN_PASSWORD
    * as I have documented directly in the .travis.yml file 
  * the package will be created automatically by the first upload from Travis
    * this will look similar to this https://bintray.com/croydon/conan/benchmark%3Agoogle
  * merge this pull request, so that the first upload is happing 
  * fill out the meta information for the package on Bintray (website, License etc.)
  * on the package overview page you can see a "Add to Conan Center" button; follow the steps 
    * conan-center is the official, central Conan repository which is pre-added for every Conan client
    * since you are the maintainers themselves of the library you will get accepted for sure 
    
    
## Result
  * no manually version string maintenance 
  * easiest way to install Benchmark for consumers
    * no git cloning
    * no manual exporting with self-defining of user/channel/version 
    * just `conan install benchmark/1.4.1@google/stable`
  * only people who are working on Benchmark itself and do install/test it via Conan, will need to specify the full reference manually
  * bonus points: you are helping conan-center to grow :)

File: conanfile.py (hunk 1 above)
Comment (Croydon) [reply conanfile.py]: Since my reply got a bit longer I answered here: https://github.com/google/benchmark/pull/647#issuecomment-408663311 😄

**Comment by LebedevRI** [comment]
Hmm wait, there is no 3rd party quality control of these packages?
And there is just one 'official' prebuilt version of the library (via travis) that is just assumed to be good enough for everybody?

**Comment by Croydon** [comment]
> Hmm wait, there is no 3rd party quality control of these packages?

There is a strict quality control when you packaging third-party libraries, it's much more relaxed when you are the maintainer of the library itself, for I guess obvious reasons.

> And there is just one 'official' prebuilt version of the library (via travis) that is just assumed to be good enough for everybody?

The Travis script I have scripted right now is not uploading prebuilt packages, only the recipe. Conan is automatically building the package if there isn't a prebuilt package available for the current os/version/arch/configuration/options...

We can surely add this as well, but the CI is then going to be a lot bigger. I wanted to start small and expend afterwards based on feedback.

**Comment by dominichamon** [comment]
is this ready for review, @Croydon ?

**Comment by Croydon** [comment]
@dominichamon Almost. I need a little bit more time to get details right. This pull request did stale for some time because I kinda wished for more early feedback 😅

Anyway, did you setup already a Bintray repository which can be used?

File: conanfile.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            del self.options.fPIC
+
+        if tools.get_env("CONAN_RUN_TESTS") != "True":
+            self.output.info("Environment variable CONAN_RUN_TEST is not defined.")
+            self.output.info("Tests are disabled and the option enable_gtest_tests isn't available.")
+            del self.options.enable_gtest_tests
+
+    def configure(self):
+        if self.settings.os == "Windows" and self.options.shared:
+            raise Exception("Windows shared builds are not supported right now, see issue #639")
```
[the remark this answers is not in the document]
Comment (Croydon) [reply conanfile.py]: Both is possible, but actually for **options** raising an exception is recommended
https://docs.conan.io/en/latest/uploading_packages/bintray/conan_center_guide.html#recipe-quality

Conan 1.8 (released today) also introduced the [ConanInvalidConfiguration exception](https://github.com/conan-io/conan/pull/3517)

**Comment by Croydon** [comment]
@dominichamon This is now ready for review. Please don't forget about setting up Bintray as it will be almost pointless without 😄

**Comment by Croydon** [comment]
The CI is looking like this on normal pushes:
https://travis-ci.org/Croydon/benchmark/builds/440290923
https://ci.appveyor.com/project/Croydon/benchmark/builds/19441354

Like this on tag pushes (in that example is 11.10.2018 the tag name):
https://travis-ci.org/Croydon/benchmark/builds/440291199
https://ci.appveyor.com/project/Croydon/benchmark/builds/19441379

And this builds the following matrix on tag pushes (= releases):
benchmarkmatrix

**Comment by dominichamon** [comment]
I've set up a benchmark repository on bintray, but i can't import this repository as it is under the 'google' organisation. I'll see what i can do to set this up.

**Comment by markgalpin** [comment]
Importing a repository primarily just sets the URL and brings in users/groups, it is not necessary to publish to bintray. You may need to set a few extra parameters manually and add the users and teams manually though.  If this is a large enough to make that impractical, let me know, and we'll work with you further.

**Comment by Croydon** [comment]
If the Bintray organisation is created and all necessary environment variables are set in both Travis and AppVeyor (as documented) then it should be good to go. The package will be automatically created by the CI via the first upload.

**Comment by Croydon** [comment]
@dominichamon Is there anything I can help with to get this over the finish line?

**Comment by dominichamon** [comment]
There are two things causing my hesitation:
1. I have no idea if this is correct because I don't know anything about Conan.
2. I haven't been able to get access to the google conan project (yet) to set up a benchmark one.

Is it safe to merge this until 2 is satisfied?

**Comment by danimtb** [comment]
Regarding point 2, I think it can be safely merged as it is not uploading anything to Bintray right now. However, the final goal is to have the packages there, so adding the variables in Travis/Appveyor that @Croydon indicated above will be the best. Thanks!

**Comment by Croydon** [comment]
> I haven't been able to get access to the google conan project (yet) to set up a benchmark one.

I have seen that you set up https://bintray.com/dominichamon/benchmark which would be already enough. The name of the package will be `benchmark/<version>@google/stable` no matter if you decide to use https://bintray.com/dominichamon/benchmark or https://bintray.com/google or another Bintray account. So I would say it's rather a question of personal taste which Bintray account you want to use.

As a reference we have done something similar with Flatbuffers https://github.com/google/flatbuffers/pull/4594

aardappel did set it up at https://bintray.com/aardappel/flatbuffers, the CI is building and upload automatically packages and Flatbuffers is included in conan-center: https://bintray.com/conan/conan-center?filterByPkgName=flatbuffers%3Agoogle

**Comment by Croydon** [comment]
> Regarding point 2, I think it can be safely merged as it is not uploading anything to Bintray right now. 

True, it is safe to merge this, but without Bintray upload it hasn't much value right now.

**Comment by dominichamon** [comment]
Help me out: Is anything else necessary for this to be checked in if we use dominichamon/benchmark (and maybe update it later)?

I'm not completely content with having API keys in the travis config in github so an alternative to that would be useful.

**Comment by Croydon** [comment]
> Help me out: Is anything else necessary for this to be checked in if we use dominichamon/benchmark (and maybe update it later)?

No, it's good to go once the Bintray access variables are set.

> I'm not completely content with having API keys in the travis config in github so an alternative to that would be useful.

Fully agreed, that it why I would recommend to set them via Travis and AppVeyor UI
For Travis: https://travis-ci.org/google/benchmark/settings
For AppVeyor: on the project page -> settings -> environment -> add environment variable

Please note that while Travis is encrypted via default, AppVeyor needs an explicit click on the lock symbol at the end of the line.


``` 
        #   CONAN_UPLOAD - a reference to the Conan repository e.g. https://api.bintray.com/conan/dominichamon/benchmark 
        #   CONAN_LOGIN_USERNAME - user for the upload e.g. dominichamon
        #   CONAN_PASSWORD - password for the CONAN_LOGIN_USERNAME; for Bintray this is the API key
```

I just noticed an integration problem with AppVeyor in general, neither for this pull request nor for the master pushes in the last half year at least (I checked with random picks) did AppVeyor run (or didn't report the status back). Maybe AppVeyor needs re-activation?
//cc @LebedevRI as LebedevRI did change the AppVeyor file the last time

**Comment by memsharded** [comment]
Hi!

Conan maintainer here.

I think this PR is very big, and I understand there are concerns about changing CI.

I would approach it step by step. The most important thing, which is not intrusive (won't change absolutely nothing in the build or in the CI) is the conanfile.py recipe and the "test_package" folder. I think that should go first, by its own.

CI is a different beast, so better wait until later. We are also working on some ideas that could help in greatly reducing or completely removing the required CI changes.

Once the conanfile.py is there, in the meantime (while we wait until the above ideas are released) we could try to setup a dedicated repo (in bincrafters or conan-community) for conan CI, without requiring to modify this repo CI scripts.

What do you think?

**Comment by dominichamon** [comment]
Hi @memsharded.

Thanks so much for weighing in! If there is a way to split this i'd really appreciate it so i can understand the changes bit-by-bit.

**Comment by Croydon** [comment]
See:
  * #728 - Minimal Conan support
  * #729 - I was mistaken, AppVeyor is enabled but doesn't report status checks; so this pull request here is completely merge ready; but the status checks should still be fixed

**Comment by vtkacenko** [comment]
@Croydon Is there anything **one** can help with to get this moving?

**Comment by dominichamon** [comment]
#728 was merged. was there anything else you were waiting for?

**Comment by vtkacenko** [comment]
@dominichamon Not waiting for anything. Just wondering if there is anything I can help with to get the https://github.com/google/benchmark/pull/647 merged; in order to have the package available in the `conan-center`.

**Comment by dominichamon** [comment]
Revisiting this. I still haven't been able to get access to bintray for google and i'm not comfortable having this hosted on my name (as it's a Google project).

Given we don't have that, it still sounds like there's a benefit to having this landed should i ever succeed in getting that access, and at _that_ point I'd need to set API keys in the various CI plugins?

**Comment by Croydon** [comment]
@dominichamon Are you on IRC/Gitter/Matrix/Slack or something similar so that we could have a direct chat about this? :)

**Comment by dominichamon** [comment]
@Croydon I'm on IRC (freenode, #googlebenchmark) most of the time.

**Comment by joachimHodara** [comment]
Thanks for your work on this issue @dominichamon and @Croydon. Any update regarding the access to bintray? Just like @vtkacenko, I would also benefit greatly from being able to get google benchmark from the conan-center. Let me know if you need help with anything.

**Comment by Croydon** [comment]
This is blocked until @dominichamon has access to the Bintray Google account or decides to use a dedicated Bintray organisation/account for hosting a Benchmark Conan repository.

The [Google Bintray](https://bintray.com/google) organisation has 5 (public) members: @tjohns @ojw28 @andrewlewis @willnorris & https://bintray.com/anaddaf

Sorry, for the seemingly random pings, but if someone of you could get in contact with @dominichamon and have a talk about getting access to this Bintray organisation this would be awesome. This issue is stalled for a long time now.

__________________

By the way, Google Flatbuffers is hosted by aardappel's personal account https://bintray.com/aardappel/flatbuffers/flatbuffers%3Agoogle

Google Fruit, on the other hand, is hosted by the Google account https://bintray.com/google/fruit

**Comment by dominichamon** [comment]
I'm still trying :)

**Comment by ojw28** [comment]
Not sure if there's some nuance here, but I don't understand why this would be difficult to sort out. @dominichamon - Feel free to email me internally if needed.

**Comment by dominichamon** [comment]
Emailed internally and followed up with other bintray owners.

I agree it shouldn't be difficult :)

**Comment by Croydon** [comment]
Ironically, 3 days after my last comment here the Conan team released the conan-center-index: https://blog.conan.io/2019/09/04/A-new-binary-building-service-for-Conan-center-packages.html

Which makes the requirement for upstream projects to set up their own repositories and CI builds go away.

I'm part of the Early Access Program and will create an inclusion request somewhen in the next days here: https://github.com/conan-io/conan-center-index

Since this is a new situation, the Conan team and community still need to figure some stuff out, for instance if upstream projects should keep the conanfiles in their own repositories as well: https://github.com/conan-io/conan-center-index/issues/22

However, the CI stuff which would be added via this pull request is not needed anymore and thereby you don't necessary need access to the Google Bintray org anymore @dominichamon

**Comment by dominichamon** [comment]
fyi, https://bintray.com/google/benchmark is there finally. let me know if there's anything i need to do to support getting this included in conan.

**Comment by Croydon** [comment]
@dominichamon [Conan Center Index](https://github.com/conan-io/conan-center-index) is really aiming to lowering the barriers for getting Conan support and ongoing maintenance. One of the perks is that it doesn't force upstream maintainers to setup their own Conan repository.

I'm really sorry about this bad timing and the work you put into getting access to the Google organisation. The Conan Center Index was just released a month ago and I didn't know any details beforehand. So I couldn't tell that it soon wouldn't be necessary anymore.

Anyhow, the goal is to get benchmark finally into the Conan Center and this should be done soon.

Pull request: https://github.com/conan-io/conan-center-index/pull/174
_______________

Within the Conan community we didn't reach any conclusion yet, [about what to do with in-source recipes](https://github.com/conan-io/conan-center-index/issues/22). Benchmark has already merged a Conan recipe in its source, which has many advantages, but maintaining stuff at two places is not well manageable right now.

As soon as we have a conclusion, I will either create a pull request in this repository here  and suggest the removal of the Conan recipe and test_package, or write a comment here and recommend to keep it :smiley:

**Comment by dominichamon** [comment]
Oh I understand completely, and I'm really happy that you're going to keep moving this forward. Wherever we end up is fine by me, at least now we have options :)

**Comment by Croydon** [comment]
Benchmark is now finally in Conan Center!

https://github.com/conan-io/conan-center-index/tree/master/recipes/benchmark
https://bintray.com/conan/conan-center/benchmark%3A_

It can be installed by either adding `benchmark/1.5.0` to a `conanfile.txt` or a `conanfile.py` file.

Or via command line: 

> conan install benchmark/1.5.0@ --build missing

Thanks anyone! 🎉 
_________________________________

I will report back once more when https://github.com/conan-io/conan-center-index/issues/22 is resolved 😄
